README.md
# Opt

Simple CLI option parsing library

## Data structures

```c
/**
 * @brief CLI root parser
 * 
 */
typedef struct cli_parser {
  /**
   * @brief Name of the CLI parser
   *
   * Can also be considered as the name of the program
   */
  char *name;

  /**
   * @brief Description of the program and what it's meant to do
   * 
   */
  char *description;

  /**
   * @brief Version of the program
   */
  char *version;
} parser_t

```

```c
/**
 * @brief Definition of the components of a command line argument
 * 
 */
typedef struct cli_option {
  /**
   * @brief Short CLI argument
   *
   * The expected format: -b
   * Normaly a single dash followed by one letter. But a couple more
   * letters are supported
   */
  const char *args;

  /**
   * @brief Long CLI argument
   *
   * The expected is format: --argname
   * The argument can be provided including the double dashes at
   * the beginning
   */
  const char *argl;

  /**
   * @brief Type of the CLI argument.
   *
   * Will help in parsing the argument
   */
  cli_option_type_t type;

  /**
   * @brief Action to do upon the reception of the argument
   *
   * Useful for boolean flags or initiate a callback function
   */
  cli_option_action_t action;

  /**
   * @brief Indicates if an argument is required
   *
   * Its default value would be '0'. When set to one,
   * the CLI parser should make sure that the argument is provided.
   */
  unsigned char required;

  /**
   * @brief Callbacl function to call when the argument is received
   * 
   */
  void (*callback)(void);

  /**
   * @brief Help message
   *
   * Reasonable length message describing the meaning of the CLI option
   */
  const char *help;

  /**
   * @brief Default value of the option
   *
   * The void pointer would be casted acording to the type of the argument
   */
  void *default_value;

  /**
   * @brief Value of the parameter after parsing
   *
   * Memory will be dynamically allocated to store the value
   */
  unsigned char *value;

} option_t;
```

```c
/**
 * @brief Type definition for parsing the CLI options
 * 
 */
typedef enum cli_option_type {
  OPT_UNKNOWN,
  OPT_BOOL,
  OPT_SHORT,
  OPT_INT,
  OPT_FLOAT,
  OPT_STR
} cli_option_type_t;
```

## API

```c
/** Create the parser */
parser_t init_parser(const char *name, const char *description);

/** Add new CLI option to the parser */
int add_arg(parser_t *parser, option_t *option);

/** Read the value of a CLI argument */
void* get_option(parser_t *parser, char *cli_arg);

/** print help */
void print_help(parser_t* parser);

/** Free all the memory dynamically allocated*/
void free_parser(parser_t *parser);
```


## Example

```c
#include <stdio.h>
#include "opt.h"

int main (int argc, char* argv[]) {
    parser_t parser = init_parser("example", "Example use of Opt library");
    // Definition of the default value of the `version` option
    char vdefault = 0;
    option_t version = {
        .args = "-v",
        .argl = "--version",
        .help = "Print out the version of the program",
        .type = OPT_BOOL,
        .default_value = &vdefault,
        .callback = print_version,
    };
    add_arg(&parser, &version);

    option_t help = {
        .args = "-h",
        .argl = "--help",
        .help = "Print this help message.",
        .type = OPT_BOOL,
    };
    add_arg(&parser, &help);

    option_t name = {
        .args = "-fn",
        .argl = "--first-name",
        .help = "First name of the user",
        .type = OPT_STR,
    };
    add_arg(&parser, &name);

    int age_default = 10;
    option_t age = {
        .args = "-age",
        .argl = "--age",
        .help = "Age of the user",
        .type = OPT_INT,
        .default_value = &age_default,
    };
    add_arg(&parser, &name);

    // Parse the CLI options
    parse(&parser, argc, argv);

    // char* vname = (char*) get_option(&parser, "-n");
    // or
    char* vname = (char*) get_option(&parser, "--name");

    int* vage = (int*) get_option(&parser, "--age");

    if (vname != NULL) printf("Name: %s\n", vname);
    else printf("CLI option '%s' not provided", name.argl);

    if (vage != NULL) printf("Age: %d\n", *vage);

    // Free parser and dynamically allocated memory
    free_parser(&parser);
    return 0;
}
```

See the `test` folder for more examples.


changelog.txt
1.0.0
    - First release.


test.c
/**
 * @file test.c
 * @author AMOUSSOU Zinsou Kenneth (www.gitlab.com/azinke)
 * @brief Testing the CLI option parsing library
 * @version 0.1
 * @date 2022-08-05
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include <stdio.h>
#include <string.h>
#include "opt.h"

#define VERSION   "0.01"

void print_version() {
  printf("test v" VERSION "\n");
}


int main (int argc, char* argv[]) {
  parser_t parser = init_parser("test", "Test program for command line argument parser");
  char vdefault = 0;
  option_t version = {
    .args = "-v",
    .argl = "--version",
    .help = "Print out the version of the program",
    .type = OPT_BOOL,
    .default_value = &vdefault,
    .callback = print_version,
  };
  add_arg(&parser, &version);

  option_t name = {
    .argl = "--name",
    .help = "Print the name of the user",
    .type = OPT_STR,
    .default_value=NULL,
  };
  add_arg(&parser, &name);

  float def_gpa = 5.0;
  option_t grade = {
    .argl = "--gpa",
    .help = "The global grade of the user",
    .type = OPT_FLOAT,
    .default_value = &def_gpa,
  };
  add_arg(&parser, &grade);

  parse(&parser, argc, argv);

  char v = *(char*)get_option(&parser, "version");
  char *vname = (char*) get_option(&parser, "name");
  float *gpa = (float*) get_option(&parser, "gpa");
  if (vname != NULL) {
    printf("Name: %s\n\n", vname);
  }
  if (gpa != NULL) {
    printf("GPA: %.8f\n\n", *gpa);
  }


  print_help(&parser);
  free_parser(&parser);
  return 0;
}


test/test_options.c
/**
 * @file test.c
 * @author AMOUSSOU Zinsou Kenneth (www.gitlab.com/azinke)
 * @brief Testing the CLI option
 * @version 0.0.1
 * @date 2022-08-05
 * 
 * @copyright Copyright (c) 2022
 * 
 */
#include <stdio.h>
#include <string.h>
#include "../opt.h"

#define VERSION   "0.0.1"

void print_version() {
  printf("test options v" VERSION "\n");
}


int main (int argc, char* argv[]) {
  parser_t parser = init_parser("test_options", "Test program for command line argument parser");
  char vdefault = 0;
  option_t version = {
    .args = "-v",
    .argl = "--version",
    .help = "Print out the version of the program",
    .type = OPT_BOOL,
    .default_value = &vdefault,
    .callback = print_version,
  };
  add_arg(&parser, &version);

  option_t name = {
    .args = "-n",
    .argl = "--name",
    .help = "Name of the user",
    .type = OPT_STR,
    .default_value=NULL,
  };
  add_arg(&parser, &name);

  float def_gpa = 5.0;
  option_t grade = {
    .argl = "--gpa",
    .help = "The global grade of the user",
    .type = OPT_FLOAT,
    .default_value = &def_gpa,
  };
  add_arg(&parser, &grade);

  parse(&parser, argc, argv);

  char v = *(char*)get_option(&parser, "--version");
  char *vname = (char*) get_option(&parser, "--name");
  float *gpa = (float*) get_option(&parser, "--gpa");
  if (vname != NULL) {
    printf("Name: %s\n\n", vname);
  }
  if (gpa != NULL) {
    printf("GPA: %.8f\n\n", *gpa);
  }

  print_help(&parser);
  free_parser(&parser);
  return 0;
}


test/test_commands.c
/**
 * @file test_commands.c
 * @author AMOUSSOU Zinsou Kenneth (www.gitlab.com/azinke)
 * @brief Test CLI commands
 * @version 0.0.1
 * @date 2022-11-05
 *
 * @copyright Copyright (c) 2022
 */

#include <stdio.h>
#include <stdlib.h>
#include "../opt.h"

#define VERSION   "0.0.1"

void print_version() {
  printf("test commands v" VERSION "\n");
  exit(0);
}

int main (int argc, char* argv[]) {
  parser_t parser = init_parser("test_commands", "CLI parser test with CRUD commands");
  char vdefault = 0;
  option_t version = {
    .args = "-v",
    .argl = "--version",
    .help = "Print out the version of the program",
    .type = OPT_BOOL,
    .default_value = &vdefault,
    .callback = print_version,
  };
  add_arg(&parser, &version);

  option_t help = {
    .args = "-h",
    .argl = "--help",
    .help = "Print this help message.",
    .type = OPT_BOOL,
  };
  add_arg(&parser, &help);

  option_t id = {
    .args = "-id",
    .help = "Item idenfification number",
    .type = OPT_STR,
  };
  add_arg(&parser, &id);

  option_t user = {
    .args = "user",
    .help = "Refer to a user. Can be combined with CRUD commands",
    .type = OPT_BOOL,
    .default_value = &vdefault,
  };
  add_arg(&parser, &user);

  option_t username = {
    .args = "-u",
    .argl = "--username",
    .help = "Screen name of the user",
    .type = OPT_STR,
  };
  add_arg(&parser, &username);

  option_t password = {
    .args = "-p",
    .argl = "--password",
    .help = "Password or secret token to access resources",
    .type = OPT_STR,
  };
  add_arg(&parser, &password);

  option_t group = {
    .args = "group",
    .help = "Refer to a group. Can be combined with CRUD commands",
    .type = OPT_BOOL,
    .default_value = &vdefault,
  };
  add_arg(&parser, &group);
  option_t name = {
    .args = "-n",
    .argl = "--name",
    .help = "Screen name of the group",
    .type = OPT_STR,
  };
  add_arg(&parser, &name);

  option_t create = {
    .args = "create",
    .help = "Create an new item",
    .type = OPT_BOOL,
    .default_value = &vdefault,
  };
  add_arg(&parser, &create);

  option_t read = {
    .args = "read",
    .help = "Read an item",
    .type = OPT_BOOL,
    .default_value = &vdefault,
  };
  add_arg(&parser, &read);

  option_t update = {
    .args = "update",
    .help = "Update an item",
    .type = OPT_BOOL,
    .default_value = &vdefault,
  };
  add_arg(&parser, &update);

  option_t delete = {
    .args = "delete",
    .help = "Delete an item",
    .type = OPT_BOOL,
    .default_value = &vdefault,
  };
  add_arg(&parser, &delete);

  parse(&parser, argc, argv);

  char* vhelp = (char*)get_option(&parser, "-h");

  if (vhelp != NULL) print_help(&parser);

  free_parser(&parser);
}


opt.c
/**
 * @file opt.c
 * @author AMOUSSOU Zinsou Kenneth (www.gitlab.com/azinke)
 * @brief CLI arguments/options parsing
 * @version 0.1
 * @date 2022-08-05
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "opt.h"


/**
 * @brief Create the CLI parser
 *
 * @param name Name of the program
 * @param description Description of the program
 * @return parser_t 
 */
parser_t init_parser(const char* name, const char* description) {
  parser_t parser = {
    .name = (char*) name,
    .description = (char*) description,
  };
  parser.first_arg = NULL;
  parser.last_arg = NULL;
  return parser;
}

/**
 * @brief Add a CLI option to a parser
 * 
 * @param parser Pointer to the target parser
 * @param option Option to be added
 * @return int 
 */
int add_arg(parser_t* parser, option_t *option) {
  // Both short and long option names cannot be NULL
  if ((option->argl == NULL) && (option->args == NULL)) {
    return EOPT_NO_ARG;
  }

  arg_t* argc = malloc(sizeof(arg_t));

  // Couldn't allocate memory
  if (argc == NULL) return EOPT_MEM_ALLOC;

  argc->next = NULL;
  argc->opt = option;
  argc->is_set = 0;

  if (parser->first_arg == NULL) {
    // Add first argument
    argc->previous = NULL;
    parser->first_arg = argc;
    parser->last_arg = argc;
  } else  {
    // Add the new argument at the end of the list
    argc->previous = parser->last_arg;
    parser->last_arg->next = (void*) argc;
    parser->last_arg = argc;
  }
  return OPT_SUCCESS;
}

/**
 * @brief Free the resources allocated to a parser
 * 
 * @param parser 
 * @return void 
 */
void free_parser(parser_t *parser) {
  while(parser->last_arg != NULL) {
    arg_t *previous = (arg_t*) parser->last_arg->previous;
    free(parser->last_arg->opt->value);
    free(parser->last_arg);
    parser->last_arg = previous;
  }
}


/**
 * @brief Print the help related to a parser
 * 
 * @param parser Pointer to the parser
 */
void print_help(parser_t* parser) {
  printf("usage: %s [command] [option]", parser->name);
  printf("\n\n%s\n\n", parser->description);
  printf("Arguments:\n");

  arg_t *argc = parser->first_arg;
  const int buffer_size = 256;
  char buf[buffer_size];
  while(argc != NULL) {
    memset(buf, '\0', buffer_size);
    if (argc->opt->args != NULL) {
      if (*(argc->opt->args) == '-') strcat(buf, "  ");
      strcat(buf, argc->opt->args);
      if (*(argc->opt->args) == '-') strcat(buf, ",");
    }
    if (argc->opt->argl != NULL) {
      if (*(argc->opt->argl) == '-') strcat(buf, "  ");
      strcat(buf, argc->opt->argl);
    }
    printf("    %-30s ", buf);
    printf("%s \n", argc->opt->help);
    argc = (arg_t*)argc->next;
  }
  printf("\n");
}


/**
 * @brief Parse the CLI arguments
 *
 * @param parser Parser
 * @param argc Argument counts
 * @param argv CLI arguments
 * @return int 
 */
int parse(parser_t *parser, int argc, char* argv[]) {
  unsigned char option_found = 0;
  unsigned char missing_value = 0;
  if (argc <= 1) return 0;
  for (int idx = 1; idx < argc; idx++) {
    arg_t *arg = parser->first_arg;
    option_found = 0;
    missing_value = 0;
    while (arg != NULL) {
      if (is_arg(argv[idx], arg) == OPT_SUCCESS) {
        option_found = 1;
        if ((arg->opt->type != OPT_BOOL) && (idx+1 >= argc)) {
          printf(
            "\033[0;31mValue expected for option '%s', "
            "but none provided.\033[0m\n\n", argv[idx]);
          missing_value = 1;
          break;
        }
        switch (arg->opt->type) {
          case OPT_BOOL: {
            // Set the boolean as "True" when present
            arg->opt->value = (unsigned char*)malloc(sizeof(char));
            *(arg->opt->value) = 1;
            break;
          }
          case OPT_SHORT: {
            arg->opt->value = (unsigned char*)malloc(sizeof(short));
            sscanf(argv[idx+1], "%hi", (short*)arg->opt->value);
            idx++; // skip the next CLI entry
            break;
          }
          case OPT_INT: {
            arg->opt->value = (unsigned char*)malloc(sizeof(int));
            sscanf(argv[idx+1], "%d", (int*)arg->opt->value);
            idx++; // skip the next CLI entry
            break;
          }
          case OPT_FLOAT: {
            arg->opt->value = (unsigned char*)malloc(sizeof(float));
            sscanf(argv[idx+1], "%f", (float*)arg->opt->value);
            idx++; // skip the next CLI entry
            break;
          }
          case OPT_STR: {
            size_t size = strlen(argv[idx+1]);
            arg->opt->value = (unsigned char*)malloc(size);
            strncpy(arg->opt->value, argv[idx+1], size);
            idx++; // skip the next CLI entry
            break;
          }
        }
        if (arg->opt->callback != NULL) {
          arg->opt->callback();
        }
        arg->is_set = 1;
        break; // Exit the while loop
      }
      arg = (arg_t*)arg->next;
    }
    if (!option_found) {
      printf("\033[0;31mUnkonwn option: '%s'\033[0m\n\n", argv[idx]);
    }
    if (missing_value || !option_found) {
      print_help(parser);
      free_parser(parser);
      exit(1);
    }
  }
}


/**
 * @brief Read the value of the CLI argument
 * 
 * @param parser 
 * @param cli_arg 
 * @return void* 
 */
void* get_option(parser_t *parser, char *cli_arg) {
  arg_t *arg = parser->first_arg;
  while (arg != NULL) {
    if (is_arg(cli_arg, arg) == OPT_SUCCESS) {
      if (arg->is_set) return arg->opt->value;
      return arg->opt->default_value;
    }
    arg = (arg_t*)arg->next;
  }
  return NULL;
}


/**
 * @brief Check if a CLI argument provided match with a predefined option
 *
 * @param cli_arg Tag of the command line argument - Should start with a single
 *                or double dash character
 * @param arg Predefined argument
 * @return int Status.
 *    EOPT_NO_ARG       : Not a CLI argument
 *    EOPT_SUCESS       : Matched found
 *    EOPT_ARG_NO_MATCH : Not matched
 */
int is_arg(char *cli_arg, arg_t *arg) {
  int arglen = strlen(cli_arg);
  int smatch = -1;
  int lmatch = -1;
  if (arg->opt->args != NULL) {
    if (strlen(arg->opt->args) == arglen) {
      smatch = strncmp(cli_arg, arg->opt->args, arglen);
    }
  }
  if (arg->opt->argl != NULL) {
    if (strlen(arg->opt->argl) == arglen) {
      lmatch = strncmp(cli_arg, arg->opt->argl, arglen);
    }
  }
  if ((smatch == 0) || (lmatch == 0)) return OPT_SUCCESS;
  return EOPT_ARG_NO_MATCH;
}


opt.h
/**
 * @file opt.h
 * @author AMOUSSOU Zinsou Kenneth (www.gitlab.com/azinke)
 * @brief CLI arguments/options parsing
 * @version 0.1
 * @date 2022-08-05
 * 
 * @copyright Copyright (c) 2022
 * 
 */

#ifndef CLI_OPT_H
#define CLI_OPT_H

#ifdef __cplusplus
extern "C" {
#endif

#define SHORT_ARG_SIZE      8
#define LONG_ARG_SIZE       16
#define HELP_ARG_SIZE       256
#define ARG_VALUE_SIZE      16


#define OPT_SUCCESS         0

#define EOPT_MEM_ALLOC      (-1)
#define EOPT_NO_ARG         (-2)
#define EOPT_ARG_EXIST      (-3)
#define EOPT_ARG_NO_MATCH   (-4)

/**
 * @brief Type definition for parsing the CLI options
 * 
 */
typedef enum cli_option_type {
  OPT_UNKNOWN,
  OPT_BOOL,
  OPT_SHORT,
  OPT_INT,
  OPT_FLOAT,
  OPT_STR
} cli_option_type_t;

/**
 * @brief Definition of CLI option actions
 *
 * Practical for boolean flags that do not require a value or executing
 * a callback function.
 */
typedef enum {
  OPT_NO_ACTION,
  OPT_STORE_TRUE,
  OPT_CALLBACK,
} cli_option_action_t;


/**
 * @brief Definition of the components of a command line argument
 * 
 */
typedef struct cli_option {
  /**
   * @brief Short CLI argument
   *
   * The expected format: -b
   * Normaly a single dash followed by one letter. But a couple more
   * letters are supported
   */
  char args[SHORT_ARG_SIZE];

  /**
   * @brief Long CLI argument
   *
   * The expected is format: --argname
   * The argument can be provided including the double dashes at
   * the beginning
   */
  char argl[LONG_ARG_SIZE];

  /**
   * @brief Type of the CLI argument.
   *
   * Will help in parsing the argument
   */
  cli_option_type_t type;

  /**
   * @brief Action to do upon the reception of the argument
   *
   * Useful for boolean flags or initiate a callback function
   */
  cli_option_action_t action;

  /**
   * @brief Callbacl function to call when the argument is received
   * 
   */
  void (*callback)(void);

  /**
   * @brief Help message
   *
   * Reasonable length message describing the meaning of the CLI option
   */
  char help[HELP_ARG_SIZE];

  /**
   * @brief Default value of the option
   *
   * The void pointer would be casted acording to the type of the argument
   */
  void *default_value;

  /**
   * @brief Value of the parameter after parsing
   *
   * Memory will be dynamically allocated to store the value
   */
  char value[ARG_VALUE_SIZE];

} option_t;


/**
 * @brief CLI Argument data structure
 *
 * The arguments will be organized in a linked list structure
 */
typedef struct cli_arg {
  /**
   * @brief CLI Argument preceding it.
   *
   * The value will be NULL of no argument comes before it 
   */
  void *previous;

  /**
   * @brief Definition of the CLI option
   */
  option_t *opt;

  /**
   * @brief Indicate of the option has been set in CLI
   *
   * If not, the default value is used.
   */
  unsigned char is_set;

  /**
   * @brief Next CLI Argument it's chained to
   * 
   */
  void *next;
} arg_t;


/**
 * @brief CLI root parser
 * 
 */
typedef struct cli_parser {

  /**
   * @brief Name of the CLI parser
   *
   * Can also be considered as the name of the program
   */
  char *name;

  /**
   * @brief Description of the program and what it's meant to do
   * 
   */
  char *description;

  /**
   * @brief Version of the program
   */
  char *version;

  /**
   * @brief Pointer to the first CLI arguments the parser is attached to
   *
   * CLI arguments will be stored in a linked list structure
   */
  arg_t *first_arg;

  /**
   * @brief Pointer to the last CLI arguments the parser is attached to
   */
  arg_t *last_arg;

} parser_t;


/** Create the parser */
parser_t init_parser(const char *name, const char *description);
void free_parser(parser_t *parser);

/** Add new CLI option to the parser */
int add_arg(parser_t *parser, option_t *option);

/** Parse arguments */
int parse(parser_t *parser, int argc, char* argv[]);

/** Read the value of a CLI argument */
void* get_option(parser_t *parser, char *cli_arg);

/** print help */
void print_help(parser_t* parser);

/** Check if a CLI argument matches one of the predefined ones */
int is_arg(char *cli_arg, arg_t *arg);


#ifdef __cplusplus
}
#endif

#endif
